[AIRFLOW-6732] Add GoogleAdsHook and GoogleAdsToGcsOperator

toddy86 opened the pull request:
---
Issue link: [AIRFLOW-6732](https://issues.apache.org/jira/browse/AIRFLOW-6732)

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

Commit:
[AIRFLOW-6732] Add GoogleAdsHook and GoogleAdsToGcsOperator

Provide ability to connect to Google Ads API using service account and execute arbitrary search (API call) using Google Ads Query language.
Convert the API results to CSV and store in designated GCS bucket.

Add google-ads to setup.py as a gcp package.

turbaszek commented on airflow/providers/google/marketing_platform/hooks/google_ads.py:
```diff
@@ -0,0 +1,151 @@
+                self._get_config()
+                self._update_config_with_secret(secrets_temp)
+                client = GoogleAdsClient.load_from_dict(self.google_ads_config)
+                return client.get_service("GoogleAdsService", version="v2")
```
Should we make the `version` parameter configurable on the hook level?

toddy86 replied:
Good point! I will update.

turbaszek commented on airflow/providers/google/marketing_platform/operators/google_ads.py:
```diff
@@ -0,0 +1,136 @@
+
+        try:
+            getter = attrgetter(*self.attributes)
+            converted_rows = [getter(row) for row in rows]
```
Nice one!

turbaszek commented on airflow/providers/google/marketing_platform/example_dags/example_google_ads.py:
```diff
@@ -0,0 +1,83 @@
+from airflow.utils import dates
+
+# [START howto_google_ads_env_variables]
+CLIENT_IDS = ["7314394221", "9725033159"]
```
Can those values be public? :)

toddy86 replied:
@nuclearpinguin They are just random number that I generated, so no concern from my end.

toddy86 replied:
I will replace with e.g. 1111111, 2222222 to expressly display that they are fake.

turbaszek reviewed (commented):
@toddy86 good job! I will add units and system this or next week

Commit:
Update operators-and-hooks-ref for Google Ads hook and operator

michalslowikowski00 commented on airflow/providers/google/marketing_platform/hooks/google_ads.py:
```diff
@@ -0,0 +1,151 @@
+    ) -> List[GoogleAdsRow]:
+        """
+        Pulls data from the Google Ads API
+
```
Should be here link to the API documentation?

toddy86 replied:
@michalslowikowski00 When I reviewed the docstrings of the other google.marketing_platform hooks I didn't note any links to API docs, so I didn't include it for consistency. But makes sense to me to have these in there, so i will add it in.

michalslowikowski00 replied:
Ok. No pressure. :)
@nuclearpinguin was a creator of marketing platform in Airflow, WDYT Tomek?

michalslowikowski00 commented on docs/howto/operator/gcp/google_ads.rst:
```diff
@@ -0,0 +1,51 @@
+    specific language governing permissions and limitations
+    under the License.
+
+Google Ads Operators
```
Should be here link to API documentation?

Commit:
Make updates for minor  PR comments

- Add api_version as an attribute of GoogleAdsHook
- Update CLIENT_IDS in example_google_ads DAG to make it clear that they are dummy values
- Add Google Ads API link in hook and operator docstrings

turbaszek committed:
Add tests

Commit:
Add flush to tempfile

Required to ensure the csvfile within the context manager is correctly flushed before being uploaded to GCS. Without this flush, the written data is not correctly saved before upload

turbaszek committed:
Fix docs

turbaszek committed:
Move code to google.ads directory

turbaszek committed:
fixup! Move code to google.ads directory

turbaszek committed:
fixup! fixup! Move code to google.ads directory

turbaszek committed:
fixup! fixup! fixup! Move code to google.ads directory

turbaszek committed:
fixup! fixup! fixup! fixup! Move code to google.ads directory
